jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java

```diff
@@ -104,15 +128,285 @@ public Object getAttribute(String name)
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (o instanceof Attributes)
+            {
+                Attributes a = (Attributes)o;
+                for (Map.Entry<String, Object> e : _map.entrySet())
+                {
+                    if (!Objects.equals(e.getValue(), a.getAttribute(e.getKey())))
+                        return false;
+                }
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * An Attributes implementation backed by another Attributes instance, which is treated as immutable, but with a
+     * ConcurrentHashMap used as a mutable layer over it.
+     */
+    class Layer implements Attributes
+    {
+        private static final Object REMOVED = "REMOVED";
+        private final Attributes _persistent;
+        private final java.util.concurrent.ConcurrentMap<String, Object> _map = new ConcurrentHashMap<>();
+        private final Set<String> _names;
+
+        public Layer(Attributes persistent)
+        {
+            _persistent = persistent;
+            _names = new LayeredNamesSet();
+        }
+
+        @Override
+        public Object removeAttribute(String name)
+        {
+            if (_persistent.getAttributeNames().contains(name))
+            {
+                Object v = _map.put(name, REMOVED);
+                if (v == REMOVED)
+                    return null;
+                if (v != null)
+                    return v;
+                return _persistent.getAttribute(name);
+            }
+            Object v = _map.remove(name);
+            return v == REMOVED ? null : v;
+        }
+
+        @Override
+        public Object setAttribute(String name, Object attribute)
+        {
+            if (attribute == null)
+                return removeAttribute(name);
+            Object v = _map.put(name, attribute);
+            return v == REMOVED ? null : v;
+        }
+
+        @Override
+        public Object getAttribute(String name)
+        {
+            Object v = _map.get(name);
+            if (v != null)
+                return v == REMOVED ? null : v;
+            return _persistent.getAttribute(name);
+        }
+
+        @Override
+        public Set<String> getAttributeNames()
+        {
+            return _names;
+        }
+
+        @Override
+        public void clearAttributes()
+        {
+            for (String n : _persistent.getAttributeNames())
+                _map.put(n, REMOVED);
+            _map.entrySet().removeIf(e -> e.getValue() != REMOVED);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            int hash = super.hashCode();
+            for (String name : getAttributeNames())
+                hash = hash ^ name.hashCode() ^ getAttribute(name).hashCode();
+            return hash;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (o instanceof Attributes)
```

Comment:
I would not write `equals` such as it can return `true` when two different implementations are used as it's too easy to break the `equals/hashcode` contract and/or have `a.equals(b) == true` while `b.equals(a) == false`.

This smells brittleness and is too easy to break in the future.

Reply:
Well at the very least, the implementation was wrong, also for hashCode!   But I'm generally OK with `Attributes` implementations being equal if they have the same contents.